src/Value.ts

```diff
@@ -0,0 +1,327 @@
+			return new FloatValue(this.value);
+		}
+
+		if (newType == ValueType.String) {
+			return new StringValue('' + this.value);
+		}
+
+		throw this.BadCastException(newType);
+	}
+}
+
+export class FloatValue extends Value<number>{
+	constructor(val: number){
+		super(val || 0.0);
+	}
+	public get isTruthy(): boolean{
+		return this.value != 0.0;
+	}
+	public get valueType(){
+		return ValueType.Float;
+	}
+
+	public Cast(newType: ValueType): Value<any>{
+		if (newType == this.valueType) {
+			return this;
+		}
+
+		if (newType == ValueType.Int) {
+			return new IntValue(this.value);
+		}
+
+		if (newType == ValueType.String) {
+			return new StringValue('' + this.value);
+		}
+
+		throw this.BadCastException(newType);
+	}
+}
+
+export class StringValue extends Value<string>{
+	public _isNewline: boolean;
+	public _isInlineWhitespace: boolean;
+
+	constructor(val: string){
+		super(val || '');
+
+		this._isNewline = (this.value == '\n');
+		this._isInlineWhitespace = true;
+
+		this.value.split('').every((c) => {
+			if (c != ' ' && c != '\t'){
+				this._isInlineWhitespace = false;
+				return false;
+			}
+
+			return true;
+		});
+	}
+	public get valueType(): ValueType{
+		return ValueType.String;
+	}
+	public get isTruthy(): boolean{
+		return this.value.length > 0;
+	}
+	public get isNewline(): boolean{
+		return this._isNewline;
+	}
+	public get isInlineWhitespace(): boolean{
+		return this._isInlineWhitespace;
+	}
+	public get isNonWhitespace(): boolean{
+		return !this.isNewline && !this.isInlineWhitespace;
+	}
+
+	public Cast(newType: ValueType): Value<any>{
+		if (newType == this.valueType) {
+			return this;
+		}
+
+		if (newType == ValueType.Int) {
+
+			let parsedInt;
+			if (parsedInt = parseInt(this.value)) {
+				return new IntValue(parsedInt);
+			} else {
+				throw this.BadCastException(newType);
+			}
+		}
+
+		if (newType == ValueType.Float) {
+			let parsedFloat;
+			if (parsedFloat = parseFloat(this.value)) {
+				return new FloatValue(parsedFloat);
+			} else {
+				throw this.BadCastException(newType);
+			}
+		}
+
+		throw this.BadCastException(newType);
+	}
+}
+
+export class DivertTargetValue extends Value<Path>{
+	constructor(targetPath: Path){
+		super(targetPath);
+
+	}
+	public get valueType(): ValueType {
+		return ValueType.DivertTarget;
+	}
+	public get targetPath(): Path{
+		return this.value;
+	}
+	public set targetPath(value: Path){
+		this.value = value;
+	}
+	public get isTruthy(): never{
+		throw new Error('Shouldn\'t be checking the truthiness of a divert target');
+	}
+
+	public Cast(newType: ValueType): Value<any>{
+		if (newType == this.valueType)
+			return this;
+
+		throw this.BadCastException(newType);
+	}
+	public toString(): string{
+		return 'DivertTargetValue(' + this.targetPath + ')';
+	}
+}
+
+export class VariablePointerValue extends Value<string>{
+	public _contextIndex: number;
+
+	constructor(variableName: string, contextIndex: number = -1){
+		super(variableName);
```

Comment:
Here, the original C# code has an empty constructor which initialize `value` to null (through super). Don't / Do we want this? (See the top comment as well.)

Reply:
Yeah I didn't change it because lazy (if `Value.value` can be `null` we need to add a bunch of safeguards), but we should stick to the C# behavior here as well.